File: flatten_gear_sets.py
# flatten_gear_sets.py
# Script to flatten nested gearSets structure in gearData.js and output as flat keys for recommendations
import re
import json

input_path = 'gearData.js'
output_path = 'gearData_flat.json'

with open(input_path, 'r', encoding='utf-8') as f:
    content = f.read()

# Extract the gearSets object (greedy match until next top-level key)
match = re.search(r'gearSets:\s*({[\s\S]*?})\s*,\s*// Gear recommendations', content)
if not match:
    print('Could not find gearSets object in gearData.js')
    exit(1)

gear_sets_str = match.group(1)

# Replace JS-style comments and trailing commas for eval safety
gear_sets_str = re.sub(r'//.*', '', gear_sets_str)
gear_sets_str = re.sub(r',\s*}', '}', gear_sets_str)
gear_sets_str = re.sub(r',\s*]', ']', gear_sets_str)

# Convert JS object keys to quoted keys for Python eval
gear_sets_str = re.sub(r'(\w+):', r'"\1":', gear_sets_str)

# Replace JS true/false/null with Python equivalents
gear_sets_str = gear_sets_str.replace('true', 'True').replace('false', 'False').replace('null', 'None')

# Evaluate to Python dict
try:
    gear_sets = eval(gear_sets_str, {"__builtins__": None}, {})
except Exception as e:
    print('Failed to parse gearSets:', e)
    exit(1)

# Flatten the nested structure
flat = {}
for class_id, roles in gear_sets.items():
    for role, sets in roles.items():
        for set_key, set_val in sets.items():
            flat_key = f"{class_id}_{role}_{set_key}"
            flat[flat_key] = set_val

# Output as JSON for easy import
with open(output_path, 'w', encoding='utf-8') as f:
    json.dump(flat, f, indent=4, ensure_ascii=False)

print(f'Flattened gearSets written to {output_path}')

File: flattenGearSets.js
// flattenGearSets.js
// Script to flatten nested gearSets structure in gearData.js and output as flat keys for recommendations

const fs = require('fs');
const path = require('path');

const gearDataPath = path.join(__dirname, 'gearData.js');
const outputPath = path.join(__dirname, 'gearData_flat.js');

// Read the original gearData.js
let fileContent = fs.readFileSync(gearDataPath, 'utf8');

// Extract the gearSets object
const gearSetsMatch = fileContent.match(/gearSets:\s*({[\s\S]*?})\s*,\s*\/\/ Gear recommendations/);
if (!gearSetsMatch) {
    console.error('Could not find gearSets object in gearData.js');
    process.exit(1);
}

const gearSetsObjStr = gearSetsMatch[1];

// Evaluate the gearSets object safely
let gearSets;
try {
    // eslint-disable-next-line no-eval
    gearSets = eval('(' + gearSetsObjStr + ')');
} catch (e) {
    console.error('Failed to parse gearSets:', e);
    process.exit(1);
}

// Flatten the nested structure
const flat = {};
for (const classId in gearSets) {
    for (const role in gearSets[classId]) {
        for (const setKey in gearSets[classId][role]) {
            // Compose flat key: classId_role_setKey
            const flatKey = `${classId}_${role}_${setKey}`;
            flat[flatKey] = gearSets[classId][role][setKey];
        }
    }
}

// Output as a recommendations object
const output = 'const recommendations = ' + JSON.stringify(flat, null, 4) + ';
\nmodule.exports = recommendations;\n';

fs.writeFileSync(outputPath, output, 'utf8');
console.log('Flattened gearSets written to gearData_flat.js');
